Kernel/Arch/riscv64/PCI/Initializer.cpp

```diff
@@ -4,17 +4,159 @@
+
+    // The pci controllers are usually in /soc/pcie?@XXXXXXXX
+    // FIXME: They can also appear in the root node, or any simple-bus other than soc
+    auto const& device_tree = DeviceTree::get();
+
+    auto maybe_soc = device_tree.get_child("soc"sv);
+    if (!maybe_soc.has_value()) {
+        dmesgln("PCI: No `soc` node found in the device tree, PCI initialization will be skipped");
+        return;
+    }
+
+    auto const& soc = maybe_soc.value();
+
+    enum class ControllerCompatible {
+        CAM,
+        ECAM,
+    };
+
+    // These properties must be present
+    auto soc_address_cells = soc.get_property("#address-cells"sv).value().as<u32>();
+    [[maybe_unused]] auto soc_size_cells = soc.get_property("#size-cells"sv).value().as<u32>();
+
+    Optional<u32> domain_counter;
+    for (auto const& entry : soc.children()) {
+        if (!entry.key.starts_with("pci"sv))
+            continue;
+        auto const& node = entry.value;
+
+        if (auto device_type = node.get_property("device_type"sv); !device_type.has_value() || device_type.value().as_string() != "pci"sv) {
+            dmesgln("PCI: PCI named device tree entry {} not a PCI controller, got device type {} instead", entry.key, device_type.has_value() ? device_type.value().as_string() : "<None>"sv);
+            continue;
+        }
+
+        auto maybe_compatible = node.get_property("compatible"sv);
+        if (!maybe_compatible.has_value()) {
+            dmesgln("PCI: PCI controller {} did not provide a compatible string, rejecting", entry.key);
+            continue;
+        }
+        auto compatible = maybe_compatible.value().as_strings();
+        ControllerCompatible controller_compatibility;
+        if (compatible.contains_slow("pci-host-ecam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::ECAM;
+        } else if (compatible.contains_slow("pci-host-cam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::CAM;
+            // FIXME: Implement CAM
+            dmesgln("PCI: PCI controller {} is CAM, which is not supported", entry.key);
+            continue;
+        } else {
+            dmesgln("PCI: PCI controller {} did not provide a compatible string, rejecting", entry.key);
+            dmesgln("PCI: Compatible strings provided: {}", compatible);
+            continue;
+        }
+
+        auto maybe_reg = node.get_property("reg"sv);
+        if (!maybe_reg.has_value()) {
+            dmesgln("PCI: PCI controller {} did not provide a physical address, rejecting", entry.key);
+            continue;
+        }
+        auto reg = maybe_reg.value();
+
+        u8 bus_range[2];
+        auto maybe_bus_range = node.get_property("bus-range"sv);
+        if (!maybe_bus_range.has_value()) {
+            bus_range[0] = 0;
+            bus_range[1] = 255;
+        } else {
+            auto bus_range = maybe_bus_range.value().as<Array<BigEndian<u32>, 2>>();
```

Comment:
Should we maybe add `as_array<T>` for integral types? So we don't have to explicitly use `BigEndian<T>` here.

Reply:
`as_array<I, N>` would either need to hand out an `Array<BigEndian<I>, N>` which is nearly invisible to use so I guess thats fine, or we need to do  the endian conversion before hand
With `Span` we can get rid of the `N` in the first case, given that it's not using `u64` (which have an alignment of 8)

Reply:
`u64`s are indeed a problem, we would have to copy them to avoid unaligned accesses.
Why should `u64`s not be a problem with `Array<BigEndian<T>, N>`?

My main complaint is that I don't like that some of the `as<T>` overloads (the integral overloads) do endian conversions. I would prefer if the endian converting function is called something else than the non converting function.
But if you think this is fine keep it like that.

Reply:
`u64`s arent an issue in `Array<BigEndian<T>, N>`, as that then has an alignment of >4 so it is returned by value,
which might be a weird API choice, I guess

Having the user not have to worry about endianness in simple cases like getting an integer seems nice to me,
thats why I like the overload for `as<I>`, but might be a thing of taste?

Reply:
Yeah, I think this API is fine. Feel free to mark as resolved.